sharedbuffers/mapped_struct.py

```diff
@@ -1719,6 +1737,374 @@ def __str__(self):
+                        not strict_subset or self.bitlen < pfset.bitlen)
+                else:
+                    dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+                    for i in xrange(xlen):
+                        if self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None) not in pfset:
+                            return False
+                    return not strict_subset or xlen < pfset.bitlen
+
+            elif self.objlist is None:
+                bitrep = self.bitrep_lo
+                while bitrep != 0:
+                    if (bitrep & 1) != 0 and i not in pfset:
+                        return False
+                    bitrep >>= 1
+                    i += 1
+
+                bitrep = self.bitrep_hi
+                while bitrep != 0:
+                    if (bitrep & 1) != 0 and i not in pfset:
+                        return False
+                    bitrep >>= 1
+                    i += 1
+
+                return not strict_subset or xlen < len(pfset)
+
+            # Both proxies use sorted lists.
+            j = 0
+            seqlen = len(pfset.objlist)
+            if xlen > seqlen:
+                return False
+
+            dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+            if dcode in ('q', 'I', 'i', 'd') and cython.compiled:
+                # fast path, use the search_hkey variants
+                dcode2, objlen2, itemsize2, offset2, _struct2 = pfset.objlist._metadata()
+                for i in xrange(xlen):
+                    val = self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None)
+                    j = pfset._search_key(val, dcode, offset, j, seqlen - j, j)
+                    if pfset.objlist._c_getitem(j, dcode2, objlen2, itemsize2, offset2, _struct2, None) != val:
+                        return False
+            else:
+                dcode2, objlen2, itemsize2, offset2, _struct2 = pfset.objlist._metadata()
+                for i in xrange(xlen):
+                    val = self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None)
+                    h1 = _stable_hash(val)
+                    while True:
+                        val2 = pfset.objlist._c_getitem(j, dcode2, objlen2, itemsize2, offset2, _struct2, None)
+                        h2 = _stable_hash(val2)
+                        j += 1
+
+                        if h1 == h2 and val == val2:
+                            break
+                        elif h1 < h2 or j >= seqlen:
+                            return False
+                        else:
+                            # pfset[j] < val, skip as much as we can
+                            while j < seqlen:
+                                if _stable_hash(pfset.objlist._c_getitem(j, dcode2, objlen2, itemsize2, offset2, _struct2, None)) >= h1:
+                                    break
+                                j += 1
```

Comment:
You should skip exponentially faster until you've skipped too much, then backtrack. As search_hkey does.